esp-hal-common/src/uart.rs

```diff
@@ -383,49 +390,59 @@ where
         }
     }
 
-    fn read_byte(&mut self) -> nb::Result<u8, Error> {
+    // As nb will be deprecated at some point, we use the async/await-style API for
+    // read_byte() internally and map it to nb-style API here
+    fn read_byte_nb(&mut self) -> nb::Result<u8, Error> {
+        match poll_once(self.read_byte()) {
+            core::task::Poll::Ready(v) => nb::Result::Ok(v),
+            core::task::Poll::Pending => nb::Result::Err(nb::Error::WouldBlock),
+        }
+    }
+
+    // Reads a single byte, blocking if needed
+    async fn read_byte(&mut self) -> u8 {
         #[allow(unused_variables)]
         let offset = 0;
 
         // on ESP32-S2 we need to use PeriBus2 to read the FIFO
         #[cfg(esp32s2)]
         let offset = 0x20c00000;
 
-        if T::get_rx_fifo_count() > 0 {
-            let value = unsafe {
-                let fifo = (T::register_block().fifo.as_ptr() as *mut u8).offset(offset)
-                    as *mut crate::peripherals::generic::Reg<FIFO_SPEC>;
-                (*fifo).read().rxfifo_rd_byte().bits()
-            };
+        while T::get_rx_fifo_count() == 0 {
+            yield_now().await;
```

Comment:
This is not good. When no data is received this will be a hot loop of yielding, meaning we can never sleep the CPU when there is nothing to do.

I'm not convinced this function should be async; if you really don't want to use the nb function (which by the way won't be deprecated, AFAIK at least) this should just be a plain old Rust function with the signature `-> Option<Result<u8, Error>>`

Reply:
I think my refactoring has not changed the behavior of the public interface at all. e.g. it was a hot loop before... But sure I can revert...

Reply:
I agree the behaviour didn't change, but the function is now marked as `async`, an unsuspecting user could call this and enter a hot loop.